BigAnimal: EHA Postgres distributions (#4181)
EnterpriseDB/docs: product_docs/docs/biganimal/release/using_cluster/01_postgres_access/pg_ba_admin.mdx

> ---

Reviewer: Scratch this one.

title: pg_ba_admin role
---

Learn about what operations the pg_ba_admin role can perform and what Postgres built-in roles it has.

## What pg_ba_admin can do

> A user with the pg_ba_admin role can perform the following operations, for example:

Reviewer: ```suggestion
A user with the pg_ba_admin role can perform many operations typically reserved for the Postgres superuser. For example:
```

- Create, modify, or delete any non-role user, including setting passwords
- Impersonate any other user using SET ROLE without requiring a password
- Grant pg_ba_admin membership and the corresponding powers to other roles
- Read contents of all Postgres system catalogs in pg_catalog
- Read and write to all user objects (tables, views, and so on)
- Perform all monitoring functions on the Postgres instance (it has the pg_monitor role)
- Install supported EDB extensions (see [Supported Postgres extensions and tools](biganimal/latest/overview/extensions_tools/))
- Bypass row level security policies
- Create foreign servers for foreign data wrappers (use may be confined by network access control rules)

## What pg_ba_admin can't do
A user with the pg_ba_admin role can't perform the following operations that a Postgres superuser can:

- Execute programs on the server (pg_ba_admin doesn't have the pg_execute_server_program role)
- Create tablespaces, casts, operator classes, access methods, and text search templates
- Read and write files on the server (pg_ba_admin doesn't have the permissions pg_read_server_files and pg_write_server_files)
- Define LEAKPROOF functions

> - Set parameters related to  pg_audit (PostgreSQL only) or edb_audit (EDB Postgres Advanced Server only)”- Create functions in non-trusted languages such as Python and C

Reviewer: ```suggestion
- Set parameters related to  pg_audit (PostgreSQL only) or edb_audit (EDB Postgres Advanced Server only)
- Create functions in non-trusted languages such as Python and C
```

- Execute CREATE LANGUAGE with custom handlers (administrators can use CREATE EXTENSION for supported languages)
- Execute CREATE FOREIGN DATA WRAPPER with custom handlers (use CREATE EXTENSION for [supported foreign data wrappers](/biganimal/release/overview/foreign_data_wrappers))
- Execute CREATE TYPE for basic types (composite types and domains are allowed)
- Execute REINDEX system catalogs
- Execute index maintenance operations of catalog tables

A user with the pg_ba_admin role can't perform the following additional operations using an EDB Postgres Advanced Server distribution:

- Use portions of the DBMS package that require writing the network or the filesystem
- Execute CREATE USER IDENTIFIED BY
- Create resource groups
- Access the network or make use of the EDB Postgres Advanced Server http packages that access the network  
- Change the encoding using DBMS_SESSION 
- Make use of the EDB Postgres Advanced Server UTIL_FILE packages that access the file system

> - Change qreplace_function setting

Reviewer: ```suggestion
- Change qreplace_function setting
- Use the DBMS_AQ package
- Use the UTL_HTTP package
- Use the UTL_TCP package
- Use the UTL_SMTP package
```

| pg_stat_scan_tables  | Execute monitoring functions that may take ACCESS SHARE locks on tables, potentially for a long time.                                                                                                                                                                                                                                              |
| pg_monitor           | Read/execute various monitoring views and functions. This role is a member of pg_read_all_settings, pg_read_all_stats and pg_stat_scan_tables.                                                                                                                                                                                                     |
| pg_database_owner    | None. Membership consists, implicitly, of the current database owner.                                                                                                                                                                                                                                                                              |

## Predefined roles not assigned to pg_ba_admin

> A user with the pg_ba_admin role doesn't has the following roles that are predefined in Postgres:

Reviewer: We need to document that these roles can't be granted by `edb_admin` either.

Author: done

> A user with the pg_ba_admin role doesn't has, and can't be granted, the following roles that are predefined in Postgres:

Reviewer: ```suggestion
A user with the pg_ba_admin role doesn't have, and can't be granted, the following roles that are predefined in Postgres:
```
